In [ ]:
import pandas as pd
import numpy as np

labels = ["label_1", "label_2", "label_3", "label_4"]
empty_label = "label_2"
features = [f'feature_{i}' for i in range(1,769)]

Importing data

In [ ]:
train_df = pd.read_csv('train.csv')
train_df.head()

feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   0.186257  -0.058807   0.024632  -0.163933  -0.146699   0.035889   
1   0.063431  -0.023597   0.068057  -0.252915  -0.061094  -0.027316   
2   0.034962   0.035816  -0.029753  -0.094607  -0.017576  -0.053074   
3   0.033772   0.085612   0.067488  -0.073953  -0.180646  -0.024512   
4   0.134305   0.062096   0.106920  -0.089327   0.117093  -0.077107   

   feature_7  feature_8  feature_9  feature_10  ...  feature_763  feature_764  \
0   0.111708  -0.162861   0.028249   -0.098063  ...     0.055629    -0.010358   
1   0.135747  -0.168147   0.091236   -0.078473  ...    -0.014893     0.071721   
2   0.040121  -0.007932   0.097872   -0.024042  ...     0.012415     0.015215   
3   0.242879  -0.023374  -0.059999    0.002006  ...    -0.078246    -0.032903   
4   0.152579   0.047529  -0.015998   -0.110657  ...    -0.094629     0.069718   

   feature_765  feature_766  feature_767  feature_768  label_1  label_2  \
0     0.125754     0.011648     0.079197     0.093215       45      NaN   
1     0.018918     0.100032    -0.083042     0.088615       45      NaN   
2     0.083808     0.031312    -0.056277     0.064702       45      NaN   
3     0.082949    -0.020659     0.082274    -0.050164       45      NaN   
4     0.014379     0.048124     0.007586    -0.016980       45      NaN   

   label_3  label_4  
0        1        6  
1        1        6  
2        1        6  
3        1        6  
4        1        6  

[5 rows x 772 columns]

In [ ]:
valid_df = pd.read_csv('valid.csv')
valid_df.head()

feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   0.085129   0.000378  -0.029872  -0.104609  -0.191291  -0.025511   
1   0.222652  -0.094210  -0.056387  -0.210214  -0.294624  -0.032355   
2   0.072147   0.023575  -0.017664  -0.070988  -0.098729   0.018534   
3   0.026623   0.052005  -0.006921  -0.075105  -0.074829  -0.031380   
4   0.112069   0.113702   0.044343   0.010162  -0.169193  -0.099847   

   feature_7  feature_8  feature_9  feature_10  ...  feature_763  feature_764  \
0   0.029381   0.006177   0.079207   -0.012729  ...    -0.026701     0.003306   
1  -0.069078  -0.172146   0.056509   -0.126757  ...    -0.041096     0.013654   
2   0.028282  -0.064789   0.094686   -0.067313  ...    -0.064602    -0.018653   
3   0.012402  -0.043555   0.073825   -0.050979  ...    -0.005655     0.015477   
4   0.195957   0.051861   0.024582    0.001314  ...    -0.118097    -0.043349   

   feature_765  feature_766  feature_767  feature_768  label_1  label_2  \
0     0.052219     0.000784     0.004527     0.102033       45      NaN   
1     0.142623    -0.011882     0.005219     0.143616       45      NaN   
2     0.013636    -0.078994     0.060903     0.064752       45      NaN   
3     0.064209     0.025045    -0.049599     0.078412       45      NaN   
4     0.036754    -0.023065     0.007358    -0.132566       45      NaN   

   label_3  label_4  
0        1        6  
1        1        6  
2        1        6  
3        1        6  
4        1        6  

[5 rows x 772 columns]

In [ ]:
test_df = pd.read_csv('test.csv')
test_df.head()

ID  feature_1  feature_2  feature_3  feature_4  feature_5  feature_6  \
0   1   0.187868  -0.009268   0.039846  -0.102515  -0.109121  -0.079769   
1   2   0.122119   0.019015   0.134483   0.007949  -0.157884  -0.033332   
2   3   0.079187   0.146195  -0.001316  -0.085596  -0.261093  -0.021620   
3   4   0.231196   0.021558   0.215534  -0.210258  -0.158189   0.041621   
4   5   0.020412   0.063732  -0.065506  -0.089598  -0.130788  -0.018809   

   feature_7  feature_8  feature_9  ...  feature_759  feature_760  \
0   0.061295  -0.042890   0.099860  ...     0.043647     0.061174   
1   0.044334  -0.016869  -0.014088  ...    -0.179798    -0.027611   
2   0.268898  -0.103950   0.068976  ...    -0.075460     0.037346   
3   0.240254  -0.068112  -0.017550  ...     0.083334     0.141642   
4   0.119304  -0.057494   0.094714  ...    -0.001969    -0.003025   

   feature_761  feature_762  feature_763  feature_764  feature_765  \
0     0.026771    -0.092734    -0.052383    -0.055526     0.029101   
1     0.240023    -0.005648    -0.057537     0.023922    -0.038126   
2     0.124254    -0.369695    -0.056773    -0.028471     0.019971   
3    -0.007219     0.031547    -0.010762    -0.169036    -0.030963   
4    -0.021156    -0.136993    -0.003615     0.036371     0.048715   

   feature_766  feature_767  feature_768  
0     0.041857    -0.027872     0.099500  
1    -0.015171    -0.006270    -0.003790  
2    -0.065683    -0.047366    -0.121744  
3     0.086698     0.114194    -0.036775  
4    -0.002688    -0.016957    -0.004331  

[5 rows x 769 columns]

In [ ]:
print(train_df.shape)
print(train_df[train_df["label_2"].notna()].shape)

print(valid_df.shape)
print(valid_df[valid_df["label_2"].notna()].shape)

(28520, 772)
(28040, 772)
(750, 772)
(736, 772)


The above result shows that label 2 has empty values in 480 and 14 rows in train and valid rows respectively. They will be handled via deletion.

**Label 2**

In [ ]:
x_train = train_df[train_df['label_2'].notna()]                                              # NaN values are dropped
x_train.drop(labels, axis=1, inplace=True)
y_train = train_df[train_df['label_2'].notna()]['label_2']

x_valid = valid_df[valid_df['label_2'].notna()]                                             # NaN values are dropped
x_valid.drop(labels, axis=1, inplace=True)
y_valid = valid_df[valid_df['label_2'].notna()]['label_2']

x_test = test_df.copy()
x_test.drop(['ID'], axis=1, inplace=True)

<ipython-input-71-0fcb4abba577>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_train.drop(labels, axis=1, inplace=True)
<ipython-input-71-0fcb4abba577>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_valid.drop(labels, axis=1, inplace=True)


Feature scaling using robust scaler

In [ ]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()
x_tr_sc = scaler.fit_transform(x_train)
x_vl_sc = scaler.fit_transform(x_valid)
x_test_sc = scaler.fit_transform(x_test)

SVM model

In [ ]:
from sklearn import svm
from sklearn.metrics import accuracy_score

svmc = svm.SVC(kernel="linear")
svmc.fit(x_tr_sc,y_train)

y_pred_svmc = svmc.predict(x_vl_sc)

acc_vl = accuracy_score(y_valid, y_pred_svmc)
acc_vl

0.8125

K-nearest model

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knnc = KNeighborsClassifier(n_neighbors=50)
knnc.fit(x_tr_sc,y_train)

y_pred_knnc = knnc.predict(x_vl_sc)

acc_vl = accuracy_score(y_valid, y_pred_knnc)
acc_vl

0.7934782608695652

Above results show that SVM classifier has a higher accuracy than k-nearest classifier with deleting of missing values.

Feature engineering with Principal Component Analysis and deletion of missing values

In [ ]:
x_train = train_df[train_df['label_2'].notna()]                                              # NaN values are dropped
x_train.drop(labels, axis=1, inplace=True)
y_train = train_df[train_df['label_2'].notna()]['label_2']

x_valid = valid_df[valid_df['label_2'].notna()]                                             # NaN values are dropped
x_valid.drop(labels, axis=1, inplace=True)
y_valid = valid_df[valid_df['label_2'].notna()]['label_2']

x_test = test_df.copy()
x_test.drop(['ID'], axis=1, inplace=True)

<ipython-input-75-0fcb4abba577>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_train.drop(labels, axis=1, inplace=True)
<ipython-input-75-0fcb4abba577>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x_valid.drop(labels, axis=1, inplace=True)


In [ ]:
from sklearn.decomposition import PCA

new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear')
  classifier.fit(x_train_trf,y_train)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 84, 103, 127, 160, 207, 278, 402]
[0.5910326086956522, 0.6005434782608695, 0.623641304347826, 0.657608695652174, 0.6834239130434783, 0.7146739130434783, 0.7432065217391305, 0.7717391304347826, 0.7948369565217391, 0.8247282608695652]
Accuracy for 50 features :- 0.5910326086956522
Accuracy for 59 features :- 0.6005434782608695
Accuracy for 70 features :- 0.623641304347826
Accuracy for 84 features :- 0.657608695652174
Accuracy for 103 features :- 0.6834239130434783
Accuracy for 127 features :- 0.7146739130434783
Accuracy for 160 features :- 0.7432065217391305
Accuracy for 207 features :- 0.7717391304347826
Accuracy for 278 features :- 0.7948369565217391
Accuracy for 402 features :- 0.8247282608695652


Using 0.98 as the n_components value;

In [ ]:
pca = PCA(n_components=0.98, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l2.csv")

In [ ]:
x_test_trf

0         1         2         3         4         5         6    \
0    0.636695  0.012724  0.695489 -0.218510 -0.192414 -0.860989  0.643005   
1    0.761557 -0.173017  1.313952 -0.525027  0.858348  0.227420  0.102094   
2   -0.474744  0.787009 -0.492960  1.376196  0.745832 -0.828283  0.558113   
3    0.404837 -0.430742  0.628836  1.900241  0.057939  1.845205 -0.637428   
4   -1.371204 -0.018001 -0.092906  0.985170 -0.013386 -0.521263  0.554698   
..        ...       ...       ...       ...       ...       ...       ...   
739 -1.313666 -1.702547  0.266016  0.009023 -1.034002 -0.143032  0.109842   
740 -0.915771 -0.250287 -0.079147  0.055090  1.079961  0.077984 -0.204223   
741 -1.423407 -1.313203  0.264768  0.347350  0.094299  0.395882  0.049371   
742 -1.118731 -0.282005  0.135730  0.038210  0.695580  0.274691 -0.434755   
743 -0.768437  1.493997  0.396134 -0.340441 -0.605914 -0.266827 -0.833087   

          7         8         9    ...       392       393       394  \
0    0.117692  0.375668 -0.375275  ...  0.014364 -0.007805  0.003104   
1    0.222636  0.200348  0.640780  ...  0.016520  0.052671  0.001635   
2   -0.355878 -0.087283  0.194659  ...  0.044028  0.007000 -0.031112   
3    0.791833  0.150910 -0.830383  ... -0.019762  0.014738  0.017482   
4   -0.322576 -0.056517  0.149672  ... -0.021813  0.003421 -0.008619   
..        ...       ...       ...  ...       ...       ...       ...   
739 -0.365876  0.324833  0.169185  ... -0.013050  0.006016 -0.027805   
740  0.250329 -0.396298 -0.216886  ... -0.002147  0.006149  0.011752   
741  0.239620 -0.059649  0.154745  ... -0.044842 -0.016453  0.038869   
742 -0.043028 -0.196309  0.050245  ... -0.001570  0.015088  0.025009   
743  0.311014 -0.374186  0.249838  ...  0.005665 -0.004493  0.039589   

          395       396       397       398       399       400       401  
0    0.005915  0.023956  0.016071  0.024317  0.022959 -0.022740  0.056244  
1    0.017187  0.019527 -0.035920 -0.018798 -0.016509  0.015530  0.022291  
2   -0.008020 -0.007665  0.037379  0.020073 -0.014293 -0.032946 -0.021194  
3    0.006845 -0.046961 -0.020062 -0.006586 -0.028801 -0.045607 -0.024864  
4    0.021240  0.001784 -0.005630 -0.021247  0.002973 -0.003876 -0.004357  
..        ...       ...       ...       ...       ...       ...       ...  
739 -0.016831 -0.034943 -0.048163 -0.002037  0.027749 -0.025170 -0.004187  
740 -0.014719  0.022392 -0.030614  0.007117 -0.006107  0.020667  0.025902  
741  0.035188  0.006651 -0.009181 -0.010183 -0.008027  0.031726 -0.003002  
742 -0.015013 -0.001266  0.013822 -0.014671 -0.008753 -0.029116  0.010137  
743  0.001873 -0.021690  0.020591  0.032992  0.019431 -0.009865  0.023977  

[744 rows x 402 columns]

In [ ]:
train_df = pd.read_csv('train.csv')
valid_df = pd.read_csv('valid.csv')
test_df = pd.read_csv('test.csv')

x_train = train_df.copy()
x_train.drop(labels, axis=1, inplace=True)
y_train_1 = train_df['label_1']
y_train_3 = train_df['label_3']
y_train_4 = train_df['label_4']

x_valid = valid_df.copy()
x_valid.drop(labels, axis=1, inplace=True)
y_valid_1 = valid_df['label_1']
y_valid_3 = valid_df['label_3']
y_valid_4 = valid_df['label_4']

x_test = test_df.copy()
x_test.drop(['ID'], axis=1, inplace=True)

**Label 1**

In [ ]:
new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear')
  classifier.fit(x_train_trf,y_train_1)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid_1,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 85, 103, 127, 161, 207, 279, 402]
[0.9053333333333333, 0.9213333333333333, 0.9373333333333334, 0.944, 0.9533333333333334, 0.9613333333333334, 0.9613333333333334, 0.9626666666666667, 0.9706666666666667, 0.972]
Accuracy for 50 features :- 0.9053333333333333
Accuracy for 59 features :- 0.9213333333333333
Accuracy for 70 features :- 0.9373333333333334
Accuracy for 85 features :- 0.944
Accuracy for 103 features :- 0.9533333333333334
Accuracy for 127 features :- 0.9613333333333334
Accuracy for 161 features :- 0.9613333333333334
Accuracy for 207 features :- 0.9626666666666667
Accuracy for 279 features :- 0.9706666666666667
Accuracy for 402 features :- 0.972


Using 0.98 as the n_components value;

In [ ]:
pca = PCA(n_components=0.98, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train_1)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l1.csv")

**Label 3**

In [ ]:
new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear')
  classifier.fit(x_train_trf,y_train_3)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid_3,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 85, 103, 127, 161, 207, 279, 402]
[0.9906666666666667, 0.9946666666666667, 0.9933333333333333, 0.996, 0.9946666666666667, 0.9986666666666667, 0.9986666666666667, 0.9986666666666667, 0.9986666666666667, 0.9986666666666667]
Accuracy for 50 features :- 0.9906666666666667
Accuracy for 59 features :- 0.9946666666666667
Accuracy for 70 features :- 0.9933333333333333
Accuracy for 85 features :- 0.996
Accuracy for 103 features :- 0.9946666666666667
Accuracy for 127 features :- 0.9986666666666667
Accuracy for 161 features :- 0.9986666666666667
Accuracy for 207 features :- 0.9986666666666667
Accuracy for 279 features :- 0.9986666666666667
Accuracy for 402 features :- 0.9986666666666667


Since accuracy stays constant after *n_components* becomes 0.9, that value, with 127 features are chosen.

In [ ]:
pca = PCA(n_components=0.9, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train_3)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l3.csv")

**Label 4**

In [ ]:
train_df['label_4'].value_counts()

6     19938
2      1449
0       955
12      954
7       938
13      482
1       481
11      480
10      480
3       479
5       478
9       472
4       469
8       465
Name: label_4, dtype: int64

Since the distribution is uneven, an additional parameter, *class_weight* with a value 'balanced' is added to the svc classifier.

In [ ]:
new_feature_numbers = []
accuracy_values = []

n_c = 0.8
while(n_c<0.99):
  pca = PCA(n_components=n_c, svd_solver='full')
  pca.fit(x_train)

  x_train_trf = pd.DataFrame(pca.transform(x_train))
  x_valid_trf = pd.DataFrame(pca.transform(x_valid))
  new_feature_numbers.append(x_train_trf.shape[1])

  classifier = svm.SVC(kernel='linear', class_weight="balanced")
  classifier.fit(x_train_trf,y_train_4)
  y_pred2 = classifier.predict(x_valid_trf)
  accuracy_values.append(accuracy_score(y_valid_4,y_pred2))
  n_c=n_c+0.02

print(new_feature_numbers)
print(accuracy_values)

i=0
while(i<len(accuracy_values)):
  print("Accuracy for "+ str(new_feature_numbers[i])+" features :- "+str(accuracy_values[i]))
  i=i+1

[50, 59, 70, 85, 103, 127, 161, 207, 279, 402]
[0.5773333333333334, 0.6266666666666667, 0.644, 0.7, 0.7333333333333333, 0.7666666666666667, 0.8, 0.8373333333333334, 0.8586666666666667, 0.8786666666666667]
Accuracy for 50 features :- 0.5773333333333334
Accuracy for 59 features :- 0.6266666666666667
Accuracy for 70 features :- 0.644
Accuracy for 85 features :- 0.7
Accuracy for 103 features :- 0.7333333333333333
Accuracy for 127 features :- 0.7666666666666667
Accuracy for 161 features :- 0.8
Accuracy for 207 features :- 0.8373333333333334
Accuracy for 279 features :- 0.8586666666666667
Accuracy for 402 features :- 0.8786666666666667


Using 0.98 as the n_components value;

In [ ]:
pca = PCA(n_components=0.98, svd_solver='full')
pca.fit(x_train)

x_train_trf = pd.DataFrame(pca.transform(x_train))
x_test_trf = pd.DataFrame(pca.transform(x_test))

classifier = svm.SVC(kernel='linear')
classifier.fit(x_train_trf,y_train_4)
y_pred = classifier.predict(x_test_trf)
pd.DataFrame(y_pred).to_csv("l4.csv")